design-proposals/sdn-ipam-secondary-nets.md

```diff
@@ -0,0 +1,450 @@
+the VM is deleted, Kubernetes Garbage Collection will kick in, and remove the
+orphaned `IPAMClaim`s. OVN-Kubernetes will need to react to the `IPAMClaim`
+delete event, to return those IP addresses to the respective IP pools.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+  participant IPAM CNI
+
+  user->>KubeVirt: deleteVM(vmName)
+
+  KubeVirt->>Kubernetes: deletePod(launcher-vmName)
+
+  Kubernetes->>CNI: CNI DEL
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+  KubeVirt-->>user: OK
+
+  loop for every orphaned IPAMClaim
+    Kubernetes->>Kubernetes: deleteIPAMClaim()
+  end
+
+  note over IPAM CNI: this CNI has an IPAMClaim informer
+
+Kubernetes->>IPAM CNI: removeEvent(ipamClaim)
+IPAM CNI->>IPAM CNI: returnToPool(ipamClaim.Status.IPs)
+```
+
+#### Hot-plug a VM interface
+This flow is exactly the same as
+[starting a VM](#starting-a-previously-stopped-virtual-machine).
+
+#### Hot-unplug a VM interface
+Hot-unplugging an interface from a KubeVirt VM is a two step process, requiring
+a migration:
+1. the user mutates the VM.Spec.Template of the running VMI setting the
+   interface(s) as absent
+2. the user migrates the VM
+3. the interfaces marked as absent will not be templated on the destination pod
+   (i.e. the migration destination pod will not have those interfaces)
+4. KubeVirt dettaches the interface from the live VM
+
+To ensure the IPs allocated to hot-unplugged interfaces are returned to the
+pool, KubeVirt needs to delete the `IPAMClaim`, otherwise, they will only be
+garbage collected when the VM is deleted.
+
+Finally, OVN-Kubernetes must react to the deletion of the IPAMClaim, and return
+this IP address to its in-memory pool.
+
+### KubeVirt related changes
+
+#### Pod templating
+
+KubeVirt would need to template the pod accordingly - i.e. the
+`NetworkSelectionElements` must feature the `ipam-claim-reference` attribute to
+indicate to the IPAM CNI / CNI which `IPAMClaim` they should use to persist the
+IP allocation across reboots / migrations.
+
+Check the following network selection elements annotation from a KubeVirt
+launcher pod to see an example:
+
+```bash
+kubectl get pods <launcher pod name> -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/networks }" | jq
+[
+  {
+    "name": "tenantred",
+    "namespace": "default",
+    "interface": "pod16367aacb67",
+    "ipam-claim-reference": "vm-a.tenantred"
+  }
+]
+```
+
+#### Creating the IPAMClaim
+KubeVirt must create the IPAMClaims (and set finalizers / owner references) to
+properly ensure the VM's lifecycle is reflected in the allocation.
+
+There are two important sections to fill when creating the IPAMClaim reference:
+- the IPAMClaim name
+- the spec.network attribute
+
+The `IPAMClaim` name will be created using the following pattern:
+`<vm name>.<logical network name>`. The logical network name is the [name of the
```

Comment:
> * I think I asked about how this is done previously, please try and add a ref to here, it may help the next reader.
> * What happens if there are two VMs with the same name in different namespaces? Is the namespace part of the IPClaim somehow?

The ipam claims are created on the namespace of the vm.

> * Having a ref to a name is problematic and raceful. Most objects have a UID to resolve the race.

The owner references of the objects carry the UID. We ensure we're handling the proper object by comparing owner references.

>   Perhaps an attempt to explain this using an example may help: We unplug and then plug quickly an interface, in the meantime the IPClaim is not yet removed completely and on creation it is still there.
>   The same can happen I guess when deleting a VM and quickly creating another one with the same name.

We check the owner references of the VMI / IPAMClaims. They must match. If they don't, we return an error, which will cause the reconcile loop to be retried later. Eventual consistency will ensure we eventually get this right (eventually the k8s GC will remove the stale resource; then eventually the "new" version manages to be created). Nothing new here IMO.

Reply:
Please add this to the text.

Regarding the potential race, I understand you use the ref owner to match the UID.
What I am not sure about is how do you manage to do the same for the hot{un}plug scenario.

Reply:
From the network interface you ref a claim, that ref is by name and not UID.
If one plugs, unplugs, plugs, unplugs etc.. the same interface, I do not understand how you can differentiate between claims of the same interface.

The difference from the VM UID is that here the interfaces are changing but the UID pointing to the VM is the same.

Reply:
I see. You're right. We do not differentiate - those would indeed be the same; the dimension is interface per VM. 

I am thinking we should do more of what we're doing: fail is there's an existing IPAMClaim, wait for it to be removed, and plug later, in a future reconcile loop. If it never gets deleted, the new interface will never be plugged into the VM.

Do you foresee any issue with this behavior ?

Reply:
Before we decide how we handle it, if by reusing or by waiting for it to be deleted
Lets discuss please the following
Is there indeed even a problem to reuse it ?
it will just have the same IP in this case

once we know the answer to that, it depends on the flow of fast hot unplug replug and dictates the easiest solution
if the above is non issue we don't need to do special care for it imho

Reply:
I think we may be able to simply indicate the dimension is interface per VM. 

As long as we adjust the user's expectation, we should be OK. 

I think it can go either way, but I feel failing until the "old" is removed is more aligned with the existing flow.

Reply:
@oshoval can you chime in ?

Reply:
We have this [logic](https://github.com/kubevirt/kubevirt/blob/100fb135bad621a9518fdcecab5cd8b6f4da41a5/pkg/virt-controller/watch/vmi.go#L1286) on VMI controller 
so if the annotation is different (previous didn't exist and now exists) at `updateMultusAnnotation` it means an interface was hotplugged, isn't it ?

Reply:
It may be raceful if the IPAMClaim was created but for some reason updating the pod annotation failed.
Maybe checkong if the claim has a deletion timestamp should be enough.